cpp/src/arrow/csv/options.h

```diff
@@ -56,6 +62,8 @@ struct ARROW_EXPORT ParseOptions {
   /// Whether empty lines are ignored.  If false, an empty line represents
   /// a single empty value (assuming a one-column CSV file).
   bool ignore_empty_lines = true;
+  /// A handler function for rows which do not have the correct number of columns
+  util::optional<InvalidRowHandler> invalid_row_handler;
```

Comment:
My vote would be that this is too much complexity.  I think simply back filling nulls (in the `InvalidRowHandler::Force` style) or skipping rows would be sufficient.  Maybe a `InvalidRowHandler` is still used under the hood to accommodate the two approaches but the interface to the user should be simpler (enum or bools).

In the rare event that someone wanted to do something custom (e.g. a unique value for each column) they could fill nulls and then in a later compute step replace null with a default value.